library/classes/Document.class.php

```diff
@@ -695,6 +845,48 @@ function createDocument(
         return '';
     }
 
+    /**
+     * Retrieves the document data that has been saved to the filesystem or couch db.  If the $decrypt flag is set to
+     * true, it will return the unencrypted version of the data for the document.
+     * @param bool $decrypt True if the document should be decrypted, false otherwise
+     * @throws BadMethodCallException Thrown if the method is called when the document has been marked as deleted or expired
+     * @return false|string Returns false if the data failed to decrypt, or a string if the data decrypts or is unencrypted.
+     */
+    function get_data($decrypt = false)
+    {
+        $storagemethod = $this->get_storagemethod();
+
+        if ($this->has_expired()) {
+            throw new BadMethodCallException("Should not attempt to retrieve data from expired documents");
+        }
+        if ($this->is_deleted()) {
+            throw new BadMethodCallException("Should not attempt to retrieve data from deleted documents");
+        }
+
+        if ($storagemethod === self::STORAGE_METHOD_COUCHDB) {
+            // Taken from ccr/display.php
+            $couch_docid = $this->get_couch_docid();
+            $couch_revid = $this->get_couch_revid();
+            $couch = new CouchDB();
+            $resp = $couch->retrieve_doc($couch_docid);
+            $data = base64_decode($resp->data);
+        } else {
+            $filepath  = $this->get_url_filepath();
```

Comment:
if only it were this easy :)
OpenEMR actually did this in the remote past, but the problem is that things then break if the instance is migrated (ie. all the paths will break). Here's an example taken from codebase (with some modes to make it work here):
```
        //change full path to current webroot.  this is for documents that may have
        //been moved from a different filesystem and the full path in the database
        //is not current.  this is also for documents that may of been moved to
        //different patients. Note that the path_depth is used to see how far down
        //the path to go. For example, originally the path_depth was always 1, which
        //only allowed things like documents/1/<file>, but now can have more structured
        //directories. For example a path_depth of 2 can give documents/encounters/1/<file>
        // etc.
        // NOTE that $from_filename and basename($url) are the same thing
        $from_all = explode("/", $filepath);
        $from_filename = array_pop($from_all);
        $from_pathname_array = array();
        for ($i = 0; $i < $this->get_path_depth(); $i++) {
            $from_pathname_array[] = array_pop($from_all);
        }
        $from_pathname_array = array_reverse($from_pathname_array);
        $from_pathname = implode("/", $from_pathname_array);
        $filepath = $GLOBALS['OE_SITE_DIR'] . '/documents/' . $from_pathname . '/' . $from_filename;
```

Reply:
Thanks for the code snippet.  I integrated it into the Document class.